### PR TITLE
feat!(db/redis): use CPE#Titles instead of CPE#Cache#Titles

### PR DESCRIPTION
**Pull request by MaineK00n**
If this Pull Request is work in progress, Add a prefix of “[WIP]” in the title.

# What did you implement:

#186 created a new key (`CPE#Cache#Titles`) when fetched, and changed to use it. 
However, there was no need to prepare `CPE#Cache#Titles`; it was enough to simply save `CPE#Titles` as a String in the first place.

## Type of change

Please delete options that are not relevant.

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# How Has This Been Tested?

# Checklist:
You don't have to satisfy all of the following.

- [ ] Write tests
- [ ] Write documentation
- [x] Check that there aren't other open pull requests for the same issue/feature
- [x] Format your source code by `make fmt`
- [x] Pass the test by `make test`
- [ ] Provide verification config / commands
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** YES 

# Reference

### CODE REVIEW

**Comment by MaineK00n** [comment]
@shino 
When should we set the deadline? (1 week ?, 1 month ?

**Comment by shino** [comment]
How about 2weeks? (geometric mean of the two)

**Comment by shino** [comment]
Diff looks nice!

One thing I don't understand yet is...
> titleListKey        = "CPE#Titles"
This key is only used in InsertCpes() . Can we remove it completely?

**Comment by MaineK00n** [comment]
@shino 
SET `CPE#Titles` cannot be deleted because it is necessary to create the String`CPE#Cache#Titles`.

**Comment by MaineK00n** [comment]
```diff
:100644 100644 6c3b7cc 0000000 M	db/redis.go

diff --git a/db/redis.go b/db/redis.go
index 6c3b7cc..bea4d6d 100644
--- a/db/redis.go
+++ b/db/redis.go
@@ -72,7 +72,6 @@ const (
 	vpSeparator         = "##"
 	deprecatedCPEsKey   = "CPE#DeprecatedCPEs"
 	titleListKey        = "CPE#Titles"
-	titleListCacheKey   = "CPE#Cache#Titles"
 	titleKeyFormat      = "CPE#Title#%s"
 	depKey              = "CPE#DEP"
 	fetchMetaKey        = "CPE#FETCHMETA"
@@ -255,31 +254,13 @@ func (r *RedisDriver) GetSimilarCpesByTitle(query string, n int, algorithm edlib
 	}
 
 	ctx := context.Background()
-	t, err := r.conn.Type(ctx, titleListKey).Result()
+	var ts []string
+	bs, err := r.conn.Get(ctx, titleListKey).Bytes()
 	if err != nil {
-		return nil, xerrors.Errorf("Failed to TYPE CPE#Titles. err: %w", err)
+		return nil, xerrors.Errorf("Failed to Get Titles. err: %w", err)
 	}
-
-	var ts []string
-	switch t {
-	case "string":
-		bs, err := r.conn.Get(ctx, titleListKey).Bytes()
-		if err != nil {
-			return nil, xerrors.Errorf("Failed to Get Titles. err: %w", err)
-		}
-		if err := json.Unmarshal(bs, &ts); err != nil {
-			return nil, xerrors.Errorf("Failed to Unmarshal JSON. err: %w", err)
-		}
-	case "set":
-		bs, err := r.conn.Get(ctx, titleListCacheKey).Bytes()
-		if err != nil {
-			return nil, xerrors.Errorf("Failed to Get Titles Cache. err: %w", err)
-		}
-		if err := json.Unmarshal(bs, &ts); err != nil {
-			return nil, xerrors.Errorf("Failed to Unmarshal JSON. err: %w", err)
-		}
-	default:
-		return nil, xerrors.Errorf("unexpected CPE#Titles type. expected: %q, actual: %q", []string{"string", "set"}, t)
+	if err := json.Unmarshal(bs, &ts); err != nil {
+		return nil, xerrors.Errorf("Failed to Unmarshal JSON. err: %w", err)
 	}
 
 	if len(ts) < n {
@@ -438,7 +419,6 @@ func (r *RedisDriver) InsertCpes(fetchType models.FetchType, cpes models.Fetched
 	for cpeURI := range oldDeps["DeprecatedCPEs"] {
 		_ = pipe.SRem(ctx, deprecatedCPEsKey, cpeURI)
 	}
-	_ = pipe.Del(ctx, titleListCacheKey)
 	for title, cpeURIs := range oldDeps["Title"] {
 		for cpeURI := range cpeURIs {
 			_ = pipe.SRem(ctx, fmt.Sprintf(titleKeyFormat, title), cpeURI)
```

File: db/redis.go [hunk 1]
```diff
@@ -379,9 +388,7 @@ func (r *RedisDriver) InsertCpes(fetchType models.FetchType, cpes models.Fetched
 						delete(oldDeps["VP"], vendorProductStr)
 					}
 				}
-				_ = pipe.SAdd(ctx, titleListKey, c.Title)
 				newDeps["Titles"][c.Title] = map[string]struct{}{}
```
Comment (shino) [inline db/redis.go:391]: "Titles" in newDeps is used in maps.Keys() below.
So, the value here can be just struct{}{} ?

----

My two cents:
The field newDeps["Titles"] is not stored in redis. 
Making a new map[string]struct{} variable "titles" (and removing "Titles" filed from newDeps) may show what is stored in redis as DEP clearly ?

File: db/redis.go [hunk 2]
```diff
@@ -256,20 +255,31 @@ func (r *RedisDriver) GetSimilarCpesByTitle(query string, n int, algorithm edlib
...
+		}
 		if err := json.Unmarshal(bs, &ts); err != nil {
 			return nil, xerrors.Errorf("Failed to Unmarshal JSON. err: %w", err)
 		}
-	} else {
-		if !xerrors.Is(err, redis.Nil) {
-			return nil, xerrors.Errorf("Failed to Get Titles. err: %w", err)
-		}
-		ts, err = r.conn.SMembers(ctx, titleListKey).Result()
+	case "set":
```
Comment (shino) [inline db/redis.go:273]: Only reading the latest source code, one can't know why this branch exists.
How about adding comment (e.g. for backward compatibility) for future source code readers (us, 99%) ?

File: db/redis.go (hunk 1 above)
Comment (MaineK00n) [reply db/redis.go:391]: > "Titles" in newDeps is used in maps.Keys() below.
So, the value here can be just struct{}{} ?

cannot use struct{}{} (value of type struct{}) as map[string]struct{} value in assignment
https://go.dev/play/p/RrdhuztnE4b

File: db/redis.go (hunk 2 above)
Comment (MaineK00n) [reply db/redis.go]: https://github.com/vulsio/go-cpe-dictionary/compare/82b51ab3fa2074c38df329f589288ccffe603835..a1b7351fdd881074f7004df4e84111619e28edbc#diff-47497c0e5aee216ce040118072f4e2199cc00ec8f2c55059b2c4729ec026e101R273

File: db/redis.go (hunk 1 above)
Comment (MaineK00n) [reply db/redis.go]: > My two cents:
The field newDeps["Titles"] is not stored in redis.
Making a new map[string]struct{} variable "titles" (and removing "Titles" filed from newDeps) may show what is stored in redis as DEP clearly ?

https://github.com/vulsio/go-cpe-dictionary/compare/82b51ab3fa2074c38df329f589288ccffe603835..a1b7351fdd881074f7004df4e84111619e28edbc#diff-47497c0e5aee216ce040118072f4e2199cc00ec8f2c55059b2c4729ec026e101R350

File: db/redis.go [hunk 3]
```diff
@@ -310,7 +327,6 @@ func (r *RedisDriver) InsertCpes(fetchType models.FetchType, cpes models.Fetched
 		"VendorProducts":           {},
 		"DeprecatedVendorProducts": {},
 		"DeprecatedCPEs":           {},
-		"Titles":                   {},
```
Comment (shino) [inline db/redis.go:313]: Remove "Titles" from the comment above "newDeps :=" line?

Comment (MaineK00n) [reply db/redis.go:313]: fixed https://github.com/vulsio/go-cpe-dictionary/compare/515af55b87e1d0e287d08824d5b5e1e69b452e0f..11a8f719f36ec2a80a1c6cd24be93dfa4d9a65c0

**Review by shino — APPROVED** [review]
fantastic!

**Pull request merged**